### PR TITLE
Implement more of the message bindings.

### PR DESCRIPTION
**Pull request by kalbasit**
Implemented:
- notmuch_message_get_replies
- notmuch_message_get_filename
- notmuch_message_get_filenames
- notmuch_message_get_date
- notmuch_message_get_header
- notmuch_message_get_tags
- notmuch_message_add_tag
- notmuch_message_remove_tag
- notmuch_message_remove_all_tags
- notmuch_message_freeze
- notmuch_message_thaw

also extends the example to show the ID, subject and the authors.

### CODE REVIEW

File: errors.go [hunk 1]
```diff
@@ -67,6 +67,10 @@ var (
 
 	// ErrUnknownError is returned when notmuch returns NULL indicating an error.
 	ErrUnknownError = errors.New("unknown error occured")
+
+	// ErrNoReplies is returned if a message has no replies or if the message's C
+	// pointer did not come from a thread.
+	ErrNoReplies = errors.New("message has no replies")
```
Comment (zenhack) [inline errors.go]: It's kind of awkward having this error be used for two different things. I understand it's not possible to tell the difference because of what the C library does, but I don't like the fact that the name reflects one of the possibilities but not the other.

File: message_test.go [hunk 2]
```diff
@@ -53,3 +58,388 @@ func TestMessageThreadID(t *testing.T) {
...
+
+	qs := "id:1258471718-6781-2-git-send-email-dottedmag@dottedmag.net"
+	threads, err := db.NewQuery(qs).Threads()
+	if err != nil {
+		t.Fatalf("error getting the threads: %s", err)
+	}
+	thread := &Thread{}
+	if !threads.Next(thread) {
+		t.Fatalf("threads.Next(thread): unable to fetch the first and only thread")
+	}
```
Comment (zenhack) [inline message_test.go:199]: This same ~10 lines of code is repeated over and over again. I'm thinking it might be prudent to wrap it in a function of its own.

File: message_test.go [hunk 3]
```diff
@@ -53,3 +58,388 @@ func TestMessageThreadID(t *testing.T) {
...
+			break
+		}
+	}
+
+	ts := msg.Tags()
+	tag := &Tag{}
+	var tags []string
+	for ts.Next(tag) {
+		tags = append(tags, tag.Value)
+	}
```
Comment (zenhack) [inline message_test.go:243]: This loop has shown up in a few different places; I'd like to see it abstracted
out. It seems like something that might even be useful to export as part of the
api; maybe a .Slice() method or something.

I'd say it feels right to put that in an add-on package, except that:
1. It feels natural to have it be a method on Tags, which we can't do
   cross-package.
2. If we want to use it in the tests, that's going to introduce a circular
   dependency.

Thoughts?

**Comment by zenhack** [comment]
I made a few comments, but this mostly looks good.

File: errors.go (hunk 1 above)
Comment (kalbasit) [inline errors.go]: I renamed it to `ErrNoRepliesOrPointerNotFromThread` a mouthful but I couldn't think of a different/shorter name.

File: message_test.go (hunk 3 above)
Comment (kalbasit) [inline message_test.go:243]: We can definitely add a `Slice()` method on `threads`, `messages` and `tags` that way, if they only thing you are doing is getting the list of threads messages or tags. It wouldn't make sense to the implemented if he calls `Slice()` only to `for range` on it again.

Not sure if an add-on package is the right thing to do here, why not simply go within our packages? Sure we are doing bindings but we are not limited to the C logic.

File: message_test.go (hunk 2 above)
Comment (kalbasit) [inline message_test.go:199]: ~~working on this, hopefully I'll get it done this weekend, sorry this week was sort-of hectic.~~ Done

**Comment by kalbasit** [comment]
@zenhack PTAL. I addressed your comments and as for the `Slice()` discussion, I feel it's out of scope for this particular PR.

File: message_test.go (hunk 3 above)
Comment (zenhack) [inline message_test.go:243]: Okay, I'm sold on it being in this package rather than an add-on. I think you're right though, we'll do it in a separate pr. I'm going to open an issue so it doesn't get forgotten.

**Comment by zenhack** [comment]
Looks good, merged.
